File: src/mustmatch/services/fixture.py
"""
Markdown document fixture for Python code blocks.

Provides an `md` object that gives Python blocks access to the
 document's structure: sections, tables, and metadata.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Iterator


def _normalize_key(value: str) -> str:
    return value.lower().replace(" ", "_").replace("-", "_")


def _coerce_value(value: str) -> Any:
    """Auto-coerce markdown table cell values to Python types."""
    stripped = value.strip()
    if stripped == "" or stripped == "None":
        return None

    lowered = stripped.lower()
    if lowered == "true":
        return True
    if lowered == "false":
        return False

    if re.fullmatch(r"-?\d+", stripped):
        return int(stripped)
    if re.fullmatch(r"-?\d+\.\d+", stripped):
        return float(stripped)

    return stripped


def _prepare_headers(headers: list[str]) -> tuple[list[str], set[str]]:
    cleaned_headers: list[str] = []
    coerce_columns: set[str] = set()

    for raw_header in headers:
        header = raw_header.strip()
        lower = header.lower()

        if lower.startswith("str:"):
            normalized = header[4:].strip()
        else:
            normalized = header
            coerce_columns.add(normalized)

        cleaned_headers.append(normalized)

    return cleaned_headers, coerce_columns


def build_table_rows(
    headers: list[str],
    row_values: list[list[str]],
) -> tuple[list[str], list[TableRow]]:
    """Build `TableRow` objects from parser table data."""
    normalized_headers, coerce_columns = _prepare_headers(headers)
    rows = [
        TableRow(
            _data=dict(zip(normalized_headers, row)),
            _coerce=coerce_columns,
        )
        for row in row_values
    ]
    return normalized_headers, rows


@dataclass
class Section:
    """A heading/section in the markdown document."""

    title: str
    level: int  # 1-6
    line: int
    parent: Section | None = None
    children: list[Section] = field(default_factory=list)


@dataclass
class TableRow:
    """A row in a table with dot-notation access to columns."""

    _data: dict[str, str]
    _coerce: set[str] = field(default_factory=set)

    def _get_key_value(self, name: str) -> tuple[str, str]:
        if name in self._data:
            return name, self._data[name]

        normalized_name = _normalize_key(name)
        for key, value in self._data.items():
            if _normalize_key(key) == normalized_name:
                return key, value

        raise KeyError(name)

    def _get_value(self, key: str, raw: str) -> Any:
        if key in self._coerce:
            return _coerce_value(raw)
        return raw

    def __getattr__(self, name: str) -> Any:
        if name.startswith("_"):
            raise AttributeError(name)

        try:
            key, raw = self._get_key_value(name)
        except KeyError as err:
            raise AttributeError(f"No column {name!r}") from err

        return self._get_value(key, raw)

    def __getitem__(self, key: str) -> Any:
        resolved_key, raw = self._get_key_value(key)
        return self._get_value(resolved_key, raw)

    def keys(self) -> list[str]:
        return list(self._data.keys())

    def values(self) -> list[Any]:
        return [self._get_value(key, raw) for key, raw in self._data.items()]

    def items(self) -> list[tuple[str, Any]]:
        return [(key, self._get_value(key, raw)) for key, raw in self._data.items()]


@dataclass
class Table:
    """A table in the markdown document."""

    name: str  # From heading context
    headers: list[str]
    rows: list[TableRow]
    line: int
    section: Section | None = None

    def __iter__(self) -> Iterator[TableRow]:
        return iter(self.rows)

    def __len__(self) -> int:
        return len(self.rows)

    def __getitem__(self, idx: int) -> TableRow:
        return self.rows[idx]

    def as_dicts(self) -> list[dict[str, Any]]:
        """Return rows as dictionaries with per-column coercion applied."""
        return [dict(row.items()) for row in self.rows]


class Tables:
    """Collection of tables with dot-notation access by name."""

    def __init__(self, tables: list[Table]) -> None:
        self._tables = tables
        self._by_name: dict[str, Table] = {}
        for table in tables:
            key = _normalize_key(table.name)
            self._by_name[key] = table

    def __getattr__(self, name: str) -> Table:
        if name.startswith("_"):
            raise AttributeError(name)
        if name in self._by_name:
            return self._by_name[name]
        raise AttributeError(f"No table {name!r}")

    def __getitem__(self, idx: int | str) -> Table:
        if isinstance(idx, int):
            return self._tables[idx]
        key = _normalize_key(idx)
        return self._by_name[key]

    def __iter__(self) -> Iterator[Table]:
        return iter(self._tables)

    def __len__(self) -> int:
        return len(self._tables)


class Sections:
    """Collection of sections with hierarchy access."""

    def __init__(self, sections: list[Section]) -> None:
        self._sections = sections
        self._by_title: dict[str, Section] = {}
        for section in sections:
            key = _normalize_key(section.title)
            self._by_title[key] = section

    def __getattr__(self, name: str) -> Section:
        if name.startswith("_"):
            raise AttributeError(name)
        if name in self._by_title:
            return self._by_title[name]
        raise AttributeError(f"No section {name!r}")

    def __getitem__(self, idx: int | str) -> Section:
        if isinstance(idx, int):
            return self._sections[idx]
        key = _normalize_key(idx)
        return self._by_title[key]

    def __iter__(self) -> Iterator[Section]:
        return iter(self._sections)

    def __len__(self) -> int:
        return len(self._sections)


@dataclass
class MD:
    """The markdown document fixture.

    Provides access to the document's structure from within Python blocks.
    """

    tables: Tables
    sections: Sections
    current_section: Section | None = None


def create_md_fixture(
    parse_result,  # ParseResult from parser
    current_block=None,  # Block being executed
) -> MD:
    """Create an MD fixture from a ParseResult.

    Args:
        parse_result: The ParseResult from parse_markdown()
        current_block: The Block currently being executed (for context)

    Returns:
        An MD object with access to document structure
    """

    sections_by_path: dict[tuple[str, ...], Section] = {}
    all_sections: list[Section] = []

    def get_or_create_section(
        *,
        level: int,
        title: str,
        path: list[str],
        line: int,
    ) -> Section:
        path_key = tuple(path)
        if path_key in sections_by_path:
            return sections_by_path[path_key]

        parent: Section | None = None
        if len(path) > 1:
            parent = sections_by_path.get(tuple(path[:-1]))

        section = Section(
            title=title,
            level=level,
            line=line,
            parent=parent,
        )
        if parent is not None:
            parent.children.append(section)

        sections_by_path[path_key] = section
        all_sections.append(section)
        return section

    for block in parse_result.blocks:
        context_lines = getattr(block, "context_lines", [])
        for idx, title in enumerate(block.context):
            get_or_create_section(
                level=idx + 1,
                title=title,
                path=block.context[: idx + 1],
                line=context_lines[idx] if idx < len(context_lines) else 0,
            )

    for parser_table in parse_result.tables:
        context_lines = getattr(parser_table, "context_lines", [])
        for idx, title in enumerate(parser_table.context):
            get_or_create_section(
                level=idx + 1,
                title=title,
                path=parser_table.context[: idx + 1],
                line=context_lines[idx] if idx < len(context_lines) else 0,
            )

    tables: list[Table] = []
    for parser_table in parse_result.tables:
        table_name = parser_table.context[-1] if parser_table.context else "unnamed"
        if parser_table.context:
            section = sections_by_path.get(tuple(parser_table.context))
        else:
            section = None

        normalized_headers, rows = build_table_rows(
            parser_table.headers,
            parser_table.rows,
        )

        tables.append(
            Table(
                name=table_name,
                headers=normalized_headers,
                rows=rows,
                line=parser_table.line_start,
                section=section,
            )
        )

    current_section = None
    if current_block and current_block.context:
        current_section = sections_by_path.get(tuple(current_block.context))

    return MD(
        tables=Tables(tables),
        sections=Sections(all_sections),
        current_section=current_section,
    )

File: crates/mustmatch-core/src/fixture.rs
use std::collections::{HashMap, HashSet};

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::coercion::{coerce_value, normalize_key};
use crate::parser::{Block, ParseResult};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SectionData {
    pub title: String,
    pub level: usize,
    pub line: usize,
    pub parent: Option<usize>,
    pub children: Vec<usize>,
    pub path: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct TableRowData {
    pub entries: Vec<(String, String)>,
    pub coerce_columns: HashSet<String>,
}

impl TableRowData {
    pub fn resolve(&self, name: &str) -> Option<(String, String)> {
        for (key, value) in &self.entries {
            if key == name {
                return Some((key.clone(), value.clone()));
            }
        }

        let normalized = normalize_key(name);
        for (key, value) in &self.entries {
            if normalize_key(key) == normalized {
                return Some((key.clone(), value.clone()));
            }
        }

        None
    }

    pub fn value_for(&self, key: &str, raw: &str) -> Value {
        if self.coerce_columns.contains(key) {
            coerce_value(raw)
        } else {
            Value::String(raw.to_string())
        }
    }

    pub fn get(&self, name: &str) -> Option<Value> {
        let (key, raw) = self.resolve(name)?;
        Some(self.value_for(&key, &raw))
    }

    pub fn keys(&self) -> Vec<String> {
        self.entries.iter().map(|(key, _)| key.clone()).collect()
    }

    pub fn values(&self) -> Vec<Value> {
        self.entries
            .iter()
            .map(|(key, raw)| self.value_for(key, raw))
            .collect()
    }

    pub fn items(&self) -> Vec<(String, Value)> {
        self.entries
            .iter()
            .map(|(key, raw)| (key.clone(), self.value_for(key, raw)))
            .collect()
    }

    pub fn data_map(&self) -> HashMap<String, String> {
        self.entries
            .iter()
            .map(|(key, value)| (key.clone(), value.clone()))
            .collect()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct FixtureTable {
    pub name: String,
    pub headers: Vec<String>,
    pub rows: Vec<TableRowData>,
    pub line: usize,
    pub section_index: Option<usize>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct MdFixture {
    pub tables: Vec<FixtureTable>,
    pub sections: Vec<SectionData>,
    pub current_section: Option<usize>,
}

fn prepare_headers(headers: &[String]) -> (Vec<String>, HashSet<String>) {
    let mut cleaned = Vec::with_capacity(headers.len());
    let mut coerce_columns = HashSet::new();

    for raw_header in headers {
        let header = raw_header.trim();
        let lower = header.to_lowercase();

        if lower.starts_with("str:") {
            cleaned.push(header[4..].trim().to_string());
        } else {
            cleaned.push(header.to_string());
            coerce_columns.insert(header.to_string());
        }
    }

    (cleaned, coerce_columns)
}

pub fn build_table_rows(
    headers: &[String],
    row_values: &[Vec<String>],
) -> (Vec<String>, Vec<TableRowData>) {
    let (normalized_headers, coerce_columns) = prepare_headers(headers);

    let rows = row_values
        .iter()
        .map(|row| {
            let entries = normalized_headers
                .iter()
                .zip(row.iter())
                .map(|(header, value)| (header.clone(), value.clone()))
                .collect();

            TableRowData {
                entries,
                coerce_columns: coerce_columns.clone(),
            }
        })
        .collect();

    (normalized_headers, rows)
}

fn path_key(path: &[String]) -> String {
    path.join("\u{001f}")
}

fn get_or_create_section(
    sections_by_path: &mut HashMap<String, usize>,
    all_sections: &mut Vec<SectionData>,
    level: usize,
    title: String,
    path: Vec<String>,
    line: usize,
) -> usize {
    let key = path_key(&path);
    if let Some(index) = sections_by_path.get(&key) {
        return *index;
    }

    let parent = if path.len() > 1 {
        let parent_key = path_key(&path[..path.len() - 1]);
        sections_by_path.get(&parent_key).copied()
    } else {
        None
    };

    let index = all_sections.len();
    all_sections.push(SectionData {
        title,
        level,
        line,
        parent,
        children: Vec::new(),
        path: path.clone(),
    });

    if let Some(parent_index) = parent {
        all_sections[parent_index].children.push(index);
    }

    sections_by_path.insert(key, index);
    index
}

pub fn create_md_fixture(parse_result: &ParseResult, current_block: Option<&Block>) -> MdFixture {
    let mut sections_by_path: HashMap<String, usize> = HashMap::new();
    let mut all_sections = Vec::new();

    for block in &parse_result.blocks {
        for (index, title) in block.context.iter().enumerate() {
            let path = block.context[..=index].to_vec();
            let line = block.context_lines.get(index).copied().unwrap_or(0);
            let _ = get_or_create_section(
                &mut sections_by_path,
                &mut all_sections,
                index + 1,
                title.clone(),
                path,
                line,
            );
        }
    }

    for table in &parse_result.tables {
        for (index, title) in table.context.iter().enumerate() {
            let path = table.context[..=index].to_vec();
            let line = table.context_lines.get(index).copied().unwrap_or(0);
            let _ = get_or_create_section(
                &mut sections_by_path,
                &mut all_sections,
                index + 1,
                title.clone(),
                path,
                line,
            );
        }
    }

    let tables = parse_result
        .tables
        .iter()
        .map(|table| {
            let name = table
                .context
                .last()
                .cloned()
                .unwrap_or_else(|| "unnamed".to_string());
            let section_index = if table.context.is_empty() {
                None
            } else {
                sections_by_path.get(&path_key(&table.context)).copied()
            };

            let (headers, rows) = build_table_rows(&table.headers, &table.rows);

            FixtureTable {
                name,
                headers,
                rows,
                line: table.line_start,
                section_index,
            }
        })
        .collect();

    let current_section = current_block.and_then(|block| {
        if block.context.is_empty() {
            None
        } else {
            sections_by_path.get(&path_key(&block.context)).copied()
        }
    });

    MdFixture {
        tables,
        sections: all_sections,
        current_section,
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use crate::parser::parse_markdown;

    use super::{build_table_rows, create_md_fixture};

    #[test]
    fn table_rows_apply_coercion_rules() {
        let headers = vec!["value".to_string(), "str:raw".to_string()];
        let rows = vec![vec!["42".to_string(), "42".to_string()]];

        let (_, built_rows) = build_table_rows(&headers, &rows);
        assert_eq!(built_rows.len(), 1);

        let row = &built_rows[0];
        assert_eq!(row.get("value"), Some(json!(42)));
        assert_eq!(row.get("raw"), Some(json!("42")));
    }

    #[test]
    fn builds_md_fixture_sections_and_current_section() {
        let source = r#"# Alpha

## Beta

| input | output |
|-------|--------|
| 2     | 4      |

```python
print("ok")
```
"#;

        let parsed = parse_markdown(source);
        let block = parsed.blocks.first().expect("one block");
        let fixture = create_md_fixture(&parsed, Some(block));

        assert_eq!(fixture.sections.len(), 2);
        assert_eq!(fixture.tables.len(), 1);
        assert!(fixture.current_section.is_some());
    }
}
